cni/network/invoker_cns.go

```diff
@@ -84,78 +85,98 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
-	if ip == nil {
-		return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-	}
+			ncgw = net.ParseIP(overlayGatewayIP)
+		}
 
-	// construct ipnet for result
-	resultIPnet := net.IPNet{
-		IP:   ip,
-		Mask: ncipnet.Mask,
-	}
+		// set result ipconfigArgument from CNS Response Body
+		ip, ncipnet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+		if ip == nil {
+			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
+		}
 
-	addResult := IPAMAddResult{}
-	addResult.ipv4Result = &cniTypesCurr.Result{
-		IPs: []*cniTypesCurr.IPConfig{
-			{
-				Address: resultIPnet,
-				Gateway: ncgw,
-			},
-		},
-		Routes: []*cniTypes.Route{
-			{
-				Dst: network.Ipv4DefaultRouteDstPrefix,
-				GW:  ncgw,
-			},
-		},
-	}
+		// construct ipnet for result
+		resultIPnet := net.IPNet{
+			IP:   ip,
+			Mask: ncipnet.Mask,
+		}
 
-	// get the name of the primary IP address
-	_, hostIPNet, err := net.ParseCIDR(info.hostSubnet)
-	if err != nil {
-		return IPAMAddResult{}, fmt.Errorf("unable to parse hostSubnet: %w", err)
-	}
+		if net.ParseIP(info.podIPAddress).To4() != nil {
+			addResult.ipv4Result = &cniTypesCurr.Result{
+				IPs: []*cniTypesCurr.IPConfig{
+					{
+						Address: resultIPnet,
+						Gateway: ncgw,
+					},
+				},
+				Routes: []*cniTypes.Route{
+					{
+						Dst: network.Ipv4DefaultRouteDstPrefix,
+						GW:  ncgw,
+					},
+				},
+			}
+		} else {
+			addResult.ipv6Result = &cniTypesCurr.Result{
+				IPs: []*cniTypesCurr.IPConfig{
+					{
+						Address: resultIPnet,
+						Gateway: ncgw,
+					},
+				},
+				Routes: []*cniTypes.Route{
+					{
+						Dst: network.Ipv4DefaultRouteDstPrefix,
+						GW:  ncgw,
+					},
+				},
+			}
+		}
 
-	addResult.hostSubnetPrefix = *hostIPNet
+		// get the name of the primary IP address
+		_, hostIPNet, err := net.ParseCIDR(info.hostSubnet)
+		if err != nil {
+			return IPAMAddResult{}, fmt.Errorf("unable to parse hostSubnet: %w", err)
+		}
 
-	// set subnet prefix for host vm
-	// setHostOptions will execute if IPAM mode is not v4 overlay
-	if invoker.ipamMode != util.V4Overlay {
-		if err := setHostOptions(ncipnet, addConfig.options, &info); err != nil {
-			return IPAMAddResult{}, err
+		addResult.hostSubnetPrefix = *hostIPNet
+
+		// set subnet prefix for host vm
+		// setHostOptions will execute if IPAM mode is not v4 overlay
+		if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualModeOverlay) {
+			if err := setHostOptions(ncipnet, addConfig.options, &info); err != nil {
+				return IPAMAddResult{}, err
+			}
 		}
 	}
 
 	return addResult, nil
 }
 
-func setHostOptions(ncSubnetPrefix *net.IPNet, options map[string]interface{}, info *IPv4ResultInfo) error {
+func setHostOptions(ncSubnetPrefix *net.IPNet, options map[string]interface{}, info *IPResultInfo) error {
```

Comment:
setHostOptions is not executed for dualstack mode. I would suggest removing these changes and add only if we need to call this in dualstack mode